controllers/vault.go

```diff
@@ -0,0 +1,281 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/go-errors/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	utilexec "k8s.io/client-go/util/exec"
+)
+
+//
+// {
+// 	"unseal_keys_b64": [
+// 	  "R/BL306DUjRQIHdkYYxheqFxr6PtZVEKtHaYNjFqBGq7",
+// 	  "+4CYavmqRWq165WJM4DqpnEqlDnECt6q+6jSmEaJBsBA",
+// 	  "tlsQ833l5k52ESK28jlZlWbegBRY+HNIJD9Yqp3cEdF6",
+// 	  "ON0wQUleo+iW4r6U0EwmoOkRhezzTke09h+rxgRDPkdo",
+// 	  "i1hmENQhAcq5t6WWxTR35YDAUjY1w8ry751CggPsB0Jk"
+// 	],
+// 	"unseal_keys_hex": [
+// 	  "47f04bdf4e83523450207764618c617aa171afa3ed65510ab4769836316a046abb",
+// 	  "fb80986af9aa456ab5eb95893380eaa6712a9439c40adeaafba8d298468906c040",
+// 	  "b65b10f37de5e64e761122b6f239599566de801458f87348243f58aa9ddc11d17a",
+// 	  "38dd3041495ea3e896e2be94d04c26a0e91185ecf34e47b4f61fabc604433e4768",
+// 	  "8b586610d42101cab9b7a596c53477e580c0523635c3caf2ef9d428203ec074264"
+// 	],
+// 	"unseal_shares": 5,
+// 	"unseal_threshold": 3,
+// 	"recovery_keys_b64": [],
+// 	"recovery_keys_hex": [],
+// 	"recovery_keys_shares": 5,
+// 	"recovery_keys_threshold": 3,
+// 	"root_token": "s.lAR1G890NPBEkzRt8Ic5kBVz"
+//   }
+// Note: We only keep the minimum needed fields around
+type VaultInitStruct struct {
+	UnsealKeysHex   []string `json:"unseal_keys_hex"`
+	UnsealShares    int      `json:"unseal_shares"`
+	UnsealThreshold int      `json:"unseal_threshold"`
+	RootToken       string   `json:"root_token"`
+}
+
+// {
+// 	"type": "shamir",
+// 	"initialized": false,
+// 	"sealed": true,
+// 	"t": 0,
+// 	"n": 0,
+// 	"progress": 0,
+// 	"nonce": "",
+// 	"version": "1.9.2",
+// 	"migration": false,
+// 	"recovery_seal": false,
+// 	"storage_type": "file",
+// 	"ha_enabled": false,
+// 	"active_time": "0001-01-01T00:00:00Z"
+//   }
+type VaultStatus struct {
+	Type        string `json:"type"`
+	Initialized bool   `json:"initialized"`
+	Sealed      bool   `json:"sealed"`
+	T           int    `json:"t"`
+	N           int    `json:"n"`
+	Progress    int    `json:"progress"`
+	Version     string `json:"version"`
+	StorageType string `json:"storage_type"`
+	HaEnabled   bool   `json:"ha_enabled"`
+}
+
+const vaultSecretName string = "vaultkeys"
+const operatorNamespace string = "patterns-operator-system"
+
+func vaultOperatorInit(config *rest.Config, client kubernetes.Interface) (*VaultInitStruct, error) {
+	stdout, stderr, err := execInPod(config, client, "vault", "vault-0", "vault", []string{"vault", "operator", "init", "-format=json"})
```

Comment:
Should `vault` and/or `vault-0` be parameters?

Reply:
So I had started playing with this and I added VaultNamespace and VaultPod to the Pattern Struct so that a user could override them, but then even if you'd override them the RBACs would start to be wrong anyways, so I am not sure it's worth it? (I mean I'll put those strings in constants so at least we avoid the repetition, just not sure allowing them being parametrized is worth it)

Reply:
that makes a lot of sense. the RBAC stuff makes this all quite painful